### ISSUE TITLE
Inventory and Items

### ISSUE THREAD

**Issue by rj00a** [valence-rs/valence#53]
The lack of inventory and items can no longer be ignored. I will be working on this next.

- [x] Item data extractor
- [x] Item code generator
- [x] Inventory packets
- [x] Inventory client events
- [ ] `inventory.rs`?
- [x] New example

**Comment by dyc3** [comment]
If we could split this out into smaller tasks, I'd like to take on some of the work.

**Comment by rj00a** [comment]
I'll see if I can do that.

**Comment by dyc3** [comment]
I've been thinking about how to structure the actual inventory types, and I think I've figured out something that might work.

Each player has an associated `PlayerInventory`:
```rust
pub struct PlayerInventory {
    armor_slots: [Slot; 4],
    offhand: Slot,
    crafting_slots: [Slot; 4],
    crafting_result: Slot,
    item_slots: Vec<Slot>,
}
```
This will allow us to keep track of the user moving items around in their inventory. The hard part is that (annoyingly) the slot numbers change depending on what type of window is open.

To account for this, we have a `WindowInventory` that will be created from the desired object's inventory, and the player's inventory. It should look something like this.

```rust
pub struct WindowInventory<T> {
    window_type: WindowType,
    object_inv: T,
    player: PlayerInventory,
}
```

Basically, this would provide a translation layer to the underlying player and object inventories. This would also mean that we _need_ to wrap all inventories in `Arc<Mutex<T>>` to make sure there aren't any data races, so really it would look like

```rust
pub struct WindowInventory<T> {
    window_type: WindowType,
    object_inv: Arc<Mutex<T>>,
    player: Arc<Mutex<PlayerInventory>>,
}
```

Reference: https://wiki.vg/Inventory

**Comment by dyc3** [comment]
Another way we could do object inventories is have it be an enum with all possible inventory types instead of having the seperate window_type field. However, I think this would make it really hard to expand valence to handle modded, non-vanilla items and inventories.

**Comment by rj00a** [comment]
I like this idea. I think you're on the right track but I have a few comments. I still don't have a complete understanding of inventories but here is what I think.

```rust
pub struct PlayerInventory {
    armor_slots: [Slot; 4],
    offhand: Slot,
    crafting_slots: [Slot; 4],
    crafting_result: Slot,
    item_slots: Vec<Slot>,
}
```

This can be represented more efficiently by using a single fixed-sized array of `Slot`, e.g. `[Slot; ?]`. This would also map more naturally to the way it's represented in the packets.

> This would also mean that we need to wrap all inventories in `Arc<Mutex<T>>` to make sure there aren't any data races, so really it would look like

Having to use an `Arc<Mutex<T>>` is usually a sign that things can be done better. Using a mutex opens the door to race conditions when the type is used outside of `Config::update` (memory-safe race conditions are still a thing). We would also have to expose the fact that a mutex is used with the entry API.

I think the correct approach is to have Valence own the inventory types, not the user. This is just like `Entity`, `Client`, `LoadedChunk` and `World` are currently. It might also make sense to have a `LoadedInventory`/`UnloadedInventory` distinction like there is with `LoadedChunk`/`UnloadedChunk`. This way, users can take ownership of `UnloadedInventory` without any issues and cannot take ownership of `LoadedInventory`.

**Comment by rj00a** [comment]
@dyc3 After having a talk with Moulberry on Discord (creator of [Graphite](https://github.com/Moulberry/Graphite)) some more pressing concerns have emerged. You can take this issue if you want.

**Comment by dyc3** [comment]
Sounds good, would you mind assigning me to the issue here?

**Comment by dyc3** [comment]
I've ran into a problem.

So, the server need to keep track of the currently opened inventory so it knows how to handle `ClickContainer` events. To be able to do this, I was just going to add a `Option<WindowInventory>` to the client.
```rust
...
open_inventory: Option<WindowInventory<I>>
...
```
The problem is that `I` needs to be known at compile time in order to work, so I it would probably be better if `WindowInventory` was not a generic.

There's 2 options I can think of:
1. Replace `I` in `WindowInventory<I>` with `dyn Inventory`, have each inventory be it's own type that implements `Inventory`. 
2. Make `Inventory` a struct (or make a new one) that lets you configure how many slots it has, the crafting slots, etc. 
In both of these scenarios, `WindowInventory` is not generic.

I'm somewhat inclined to go with option 2 for the sake of compile times.

**Comment by dyc3** [comment]
Now that I think about it, option 2 might make it really clunky to implement custom menus.

**Comment by dyc3** [comment]
Now that I think about it, option 2 might make it really clunky to implement custom menus? I'll have to try it out.

**Comment by dyc3** [comment]
In the draft PR, I've created an example that shows WindowInventory being used to modify a chest's inventory. It's functional, but it works by manually sending packets.

There's a few problems to solve/decide before figuring out what the user facing API going to look like:
1. State synchronization. This is the big one. If there is a change to an inventory, how do we propagate that change to all subscribed clients transparently? It should Just Work™. Multiple players must be able to open a chest and see it being changed.
2. How do we let users change the behavior of an inventory?
3. Do we allow users to directly access inventories? State sync is important here.

### State Synchronization
Here's what I'm thinking:

Each non-player, non-window Inventory has a list of subscribers. For brevity, lets call that a "chest-like" inventory. Whenever a chest-like is changed, it sends the update to the `WindowInventory`s that are subscribed. The `WindowInventory` objects are associated with a single client, so they handle sending that update to the client.

Next problem, we need some way to be able to mark an inventory or individual slots as dirty so that we can know to publish updates to subscribers. There's 2 ways to do this:
- Just have getter/setter methods where we can set the dirty flag
- Have something like `MutexGuard` that impls `DerefMut` so we can set the dirty flag whenever there is a mutable borrow.

This would probably mean that all inventories need to be managed by Valence. (There's definitely more that needs to go into this, I might edit this later with more details.)

**Comment by rj00a** [comment]
Would it be better if the clients owned the list of inventories they were subscribed to (list of `InventoryId` or something)? Then at the end of each tick the clients would check the inventories for modifications and apply the changes there. After client updates the inventory modifications can be cleared. And yea this means inventories would be managed by Valence.

I haven't taken a closer look yet.

**Comment by dyc3** [comment]
Maybe? Idk, I'll have to try it out.

**Comment by dyc3** [comment]
I've updated the draft PR with state synchronization. There are some race condition problems with the `chest` example that I have to work out. If you try to take an item out of the chest, it's more likely than not to duplicate the item.
